### PR TITLE
Support automatic Let's Encrypt

### LINKED ISSUE #105: Support automatic Let's Encrypt certificate aquisition
### Proposed change

Add an option to automatically get let's encrypt certificates for a given set of domains. Traefik already supports this, and we should expose this easily.

### Alternative options

Alternatives are:

1. We add extra traefik config to get our TLS certs (we do this in TLJH)
2. Add an additional proxy in front for TLS termination (we do this in z2jh, although not with traefik proxy)

It would be nice to not have to do either of these things!

### Who would use this feature?

Everyone, since everyone wantsn HTTPS!

### (Optional): Suggest a solution

Upstream the automatic HTTPS code from TLJH to traefik-proxy

### PR DESCRIPTION
**Pull request by GeorgianaElena**
This PR adds:
- [x] Support configuring acme when proxy is started by the Hub
   - http challenge only
- [x] Auto https test
  - using [pebble](https://github.com/letsencrypt/pebble) (thanks @consideRatio for helping out with this :heart: )
- [ ] Update docs

Closes #105

### CODE REVIEW

File: jupyterhub_traefik_proxy/toml.py [hunk 1]
```diff
@@ -94,6 +94,8 @@ def _clean_resources(self):
         try:
             if self.should_start:
                 os.remove(self.toml_static_config_file)
+                if self.traefik_auto_https:
+                    os.remove(self.traefik_acme_storage)
```
Comment (consideRatio) [inline jupyterhub_traefik_proxy/toml.py:98]: It seems like _clean_resources is called along with .stop() which "Will be called during teardown if should_start is True.".

So, it seems then to me, that this configuration guarantees that everything in this ACME storage will reset on start/stop cycles. I think that this is where the acquired certificates will reside, then they should probably only be removed if we really want that specifically, but if we want to just restart, we probably want to reuse our previously acquired certificates.

File: tests/conftest.py [hunk 2]
```diff
@@ -13,6 +13,41 @@
 from jupyterhub_traefik_proxy import TraefikTomlProxy
 
 
+@pytest.fixture
+async def autohttps_toml_proxy():
+    """Fixture returning a configured Let's Encrypt TraefikTomlProxy"""
+    proxy = TraefikTomlProxy(
+        public_url="http://127.0.0.1:8000",
```
Comment (consideRatio) [inline tests/conftest.py:20]: I'm not sure about what to test, but perhaps this should be `https://jupyter.test:8443`?

File: tests/proxytest.py [hunk 3]
```diff
@@ -356,3 +356,36 @@ async def test_websockets(proxy, launch_backend):
...
+        "https://"
+        + urlparse(proxy.public_url).hostname
+        + ":"
+        + str(proxy.traefik_https_port)
+        + routespec
+    )
+    assert resp.effective_url == expected_final_redirect_url
+
+    # Test redirection to the route added
+    assert backend_response == backend_port
```
Comment (consideRatio) [inline tests/proxytest.py:391]: Are we only testing redirection in this test?

It would perhaps make sense to test a successful response to https as well perhaps, than proves that TLS termination (encrypt/decrypt etc) is done successfully. The client that makes this webrequest would need to not try verify the TLS certificate (`verify=False` in python's requests library, as passed with the `get`, `post`, and etc functions).

An issue I experienced with z2jh was that I had to await traefik being ready, and found that awaiting the logs from Pebble saying it has issued a certificate was good enough for my test suite to ensure my traefik instance was ready, while it wasn't enough for me to await the log entry "Registering certificate..." made by traefik - i think 2-5 seconds additional to that entry showing was needed for traefiks HTTPS endpoints to be ready.

File: docs/source/https.md [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    ```python
+    c.Proxy.traefik_auto_https=True
+    ```
+    * Configure a HTTPS port. This is where all the traffic coming through the HTTP entrypoint will be redirected to:
+    ```python
+    c.Proxy.traefik_https_port=443
+    ```
+    * Set the email address used for Let's Encrypt registration:
+    ```python
+    c.Proxy.traefik_letsencrypt_email="jovyan@jupyter.test"
```
Comment (consideRatio) [inline docs/source/https.md:20]: I suggest we leave this blank, and let it fail if it is and it is needed. That way, we don't risk having people using this code as it and sending requests with user accounts associated with Jupyter which is not associated with an actual human.

```suggestion
    c.Proxy.traefik_letsencrypt_email=""
```

File: jupyterhub_traefik_proxy/proxy.py [hunk 5]
```diff
@@ -47,15 +47,29 @@ class TraefikProxy(Proxy):
...
+
+    traefik_auto_https = Bool(
+        False, config=True, help="""enable automatic HTTPS with Let's Encrypt"""
+    )
+
+    traefik_letsencrypt_email = Unicode(config=True, help="""Let's Encrypt email""")
+
+    traefik_letsencrypt_domains = List(config=True, help="""domains list""")
+
+    traefik_acme_server = Unicode(config=True, help="""the CA server to use""")
```
Comment (minrk) [inline jupyterhub_traefik_proxy/proxy.py:65]: Does it make sense for the default acme server to be letsencrypt staging? `https://acme-staging-v02.api.letsencrypt.org/directory`

That way it works while you are testing without configuring the production server, plus we encourage folks to start using staging, which is a good idea anyway, since it avoids rate limit issues while folks redo initial deployments.

File: jupyterhub_traefik_proxy/proxy.py [hunk 6]
```diff
@@ -217,21 +231,49 @@ async def _setup_traefik_static_config(self):
         self.static_config = {}
         self.static_config["debug"] = True
         self.static_config["logLevel"] = self.traefik_log_level
+
         entryPoints = {}
+        self.static_config["defaultentrypoints"] = ["http"]
+        entryPoints["http"] = {"address": ":" + str(urlparse(self.public_url).port)}
```
Comment (minrk) [inline jupyterhub_traefik_proxy/proxy.py:237]: this should respect the ip, if given, of public_url. Maybe use `urlparse(self.public_url).netloc`? What happens if public_url has the form `https://1.2.3.4` which has an implicit port of 443 (inferred from https), but the parsed `url.port` will be None.

File: docs/source/https.md [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    Providing a certificate and key can be done by configuring JupyterHub to enable SSL encryption as specified in [the docs](https://jupyterhub.readthedocs.io/en/stable/getting-started/security-basics.html?highlight=https#enabling-ssl-encryption). Example:
+    ```python
+      c.JupyterHub.ssl_key = '/path/to/my.key'
+      c.JupyterHub.ssl_cert = '/path/to/my.cert'
+    ```
+
+    Traefik will then redirect all HTTP traffic to the HTTPS entrypoint. The default HTTPS entrypoint port is 8843, but can be configured through:
+    ```python
+    c.Proxy.traefik_https_port=443
+    ```
```
Comment (consideRatio) [inline docs/source/https.md:47]: It is fine to describe this configuration twice, just wanted to make you aware of that is done and you have the option to avoid writing it two times.

File: jupyterhub_traefik_proxy/proxy.py (hunk 6 above)
Comment (GeorgianaElena) [reply jupyterhub_traefik_proxy/proxy.py:237]: Oh, right! I didn't thought about this case.
Should we just then have one entrypoint `http` or `https` inferred from `public_url`? (I think this is how things happen in CHP)

Right now, (in this form of the PR), there is:
* the http entrypoint - that's setup through `public_url`
* the https entrypoint - that's setup through `traefik_https_port`
And all the requests coming for http are redirected to https.

File: tests/conftest.py (hunk 2 above)
Comment (GeorgianaElena) [reply tests/conftest.py:20]: I think you're right (checkout [this comment](https://github.com/jupyterhub/traefik-proxy/pull/108#discussion_r427219235) for what was going on in my head about `public_url` :sweat_smile:)

File: tests/proxytest.py (hunk 3 above)
Comment (GeorgianaElena) [reply tests/proxytest.py:391]: > Are we only testing redirection in this test?

The test should follow this flow:
* launch a [http server on 9900](https://github.com/GeorgianaElena/traefik-proxy/blob/auto_https/tests/proxytest.py#L368) (that only knows to respond to GET request with the port is running on).
* [wait for traefik and the http server](https://github.com/GeorgianaElena/traefik-proxy/blob/auto_https/tests/proxytest.py#L370) show up on their specified ports
* [add a route to traefik](https://github.com/GeorgianaElena/traefik-proxy/blob/auto_https/tests/proxytest.py#L372) from `/autohttps` to the backend http server
* make [a http request](https://github.com/GeorgianaElena/traefik-proxy/blob/auto_https/tests/proxytest.py#L372) for the public_url and the route we just added.
The request should be redirected as:
`http://localhost:8000/autohttps` -> (because we're redirecting traffic from the http entrypoint on port 8000 to https) 
`https://localhost:8443/autohttps` -> (because we added the route to the backend dummy server)
`http://localhost:9900` that should respond with a 200 code and `9900` as the payload.

I thought that testing this redirection will mean that we also got a successful response to https :confused: 

> The client that makes this webrequest would need to not try verify the TLS certificate (verify=False in python's requests library, as passed with the get, post, and etc functions)

The request has the validate_cert flag set to False [here](https://github.com/GeorgianaElena/traefik-proxy/blob/auto_https/tests/proxytest.py#L375).

> An issue I experienced with z2jh was that I had to await traefik being ready, and found that awaiting the logs from Pebble saying it has issued a certificate was good enough for my test suite to ensure my traefik instance was ready, while it wasn't enough for me to await the log entry "Registering certificate..." made by traefik - i think 2-5 seconds additional to that entry showing was needed for traefiks HTTPS endpoints to be ready.

I noticed that registering the certificate takes a while. From the logs, it looks that the request seems to happen after the certificate acquisition. I'm wondering if it's just out of luck or [this check](https://github.com/GeorgianaElena/traefik-proxy/blob/auto_https/test/proxytest.py#L370) is enough or maybe the certificate acquisition is blocking (I didn't find any evidence of this in traefik's code to support this though).

Comment (GeorgianaElena) [reply tests/proxytest.py:391]: It looks like the test is indeed not testing the right thing: https://travis-ci.org/github/GeorgianaElena/traefik-proxy/jobs/688791781 passes even though it cannot connect to the acme client :confused:

File: docs/source/https.md (hunk 7 above)
Comment (GeorgianaElena) [reply docs/source/https.md:47]: I thought it would be a good idea to have it here also for the cases when users just come to this section from the table of contents and don't necessarily read what's in the section above because that isn't the use case they are looking for.

File: docs/source/https.md [hunk 8]
```diff
@@ -27,6 +23,10 @@ up **manually** with your own key and certificate.
     ```python
     c.Proxy.traefik_acme_server="https://acme-v02.api.letsencrypt.org/directory"
     ```
+    * Set the port to be used by Traefik for the Acme HTTP challenge:
+    ```python
+    c.Proxy.traefik_acme_challenge_port="https://acme-v02.api.letsencrypt.org/directory"
```
Comment (consideRatio) [inline docs/source/https.md:28]: A blooper =)

Comment (GeorgianaElena) [reply docs/source/https.md:28]: Hehe :smile: copy-paste gone wrong. Thanks.

Comment (consideRatio) [inline docs/source/https.md]: Hmmm, Traefik does indeed need to listen to HTTP based traffic, so that it can receive traffic from an ACME server sending out a HTTP-01 request to port 80. I'm not sure how the configuration is setup etc, but perhaps you can rephrase this to indicate that its the port where traefik will be _listening_ to a HTTP-01 challenge request from an ACME server, which only sends out request to port 80 (except for Pebble - the ACME testing server).

File: docs/source/https.md [hunk 9]
```diff
@@ -25,7 +25,8 @@ up **manually** with your own key and certificate.
     ```
     * Set the port to be used by Traefik for the Acme HTTP challenge:
     ```python
-    c.Proxy.traefik_acme_challenge_port="https://acme-v02.api.letsencrypt.org/directory"
+    # default port is 80
```
Comment (consideRatio) [inline docs/source/https.md:28]: There are too many ports that could be referenced implicitly, be more explicit, the ACME server sends out its HTTP-01 challenges to port 80 by default, but is the default value of this configuration port 80?

Comment (GeorgianaElena) [reply docs/source/https.md:28]: Maybe is better if I remove this step entirely from the instructions and a note at the bottom about the need to change this port only when testing using Pebble?

Comment (consideRatio) [reply docs/source/https.md:28]: Sounds fine to me!

File: .travis.yml [hunk 10]
```diff
@@ -1,15 +1,29 @@
 language: python
 cache:
   - pip
+addons:
+  hosts:
+    - jupyter.test
```
Comment (consideRatio) [inline .travis.yml:6]: I learned recently from Min that we could use local.jovyan.org instead to point to localhost, and by doing so avoid the need to setup things in /etc/hosts or like we do here for travis.

Suggestion: replace jupyter.test with local.jovyan.org

File: tests/conftest.py [hunk 11]
```diff
@@ -6,13 +6,36 @@
...
+    """Fixture returning a configured Let's Encrypt TraefikTomlProxy"""
+    proxy = TraefikTomlProxy(
+        public_url="https://jupyter.test:8443",
+        traefik_api_password="admin",
+        traefik_api_username="api_admin",
+        should_start=True,
+        traefik_auto_https=True,
+        traefik_letsencrypt_email="jovyan@jupyter.test",
+        traefik_letsencrypt_domains=["jupyter.test"],
+        traefik_acme_server="https://0.0.0.0:14000/dir",
```
Comment (consideRatio) [inline tests/conftest.py:28]: Is this correct? Is this the configuration of where traefik is supposed to find the ACME server? Then we need an actual address I think. I think if we have started Pebble locally, then we should write `localhost` here. We could write 127.0.0.1 but then our host header with the request would be 127.0.0.1 instead of localhost, and that is bad for the TLS certificate pebble will have which allows for `localhost` but not `127.0.0.1` I think.

Comment (GeorgianaElena) [reply tests/conftest.py:28]: From the logs, it looks like the interaction between traefik and the acme server happens ok and they find each other, but I'll switch to localhost to give it an actual address and be specific.

File: .travis.yml (hunk 10 above)
Comment (GeorgianaElena) [reply .travis.yml:6]: That's really cool :sparkles: Thanks @consideRatio :heart:

**Comment by GeorgianaElena** [comment]
@minrk, do you think there is anything else I should change/improve before merging this?
(sorry for the ping)
